google/cloud/logging_v2/handlers/_monitored_resources.py

```diff
@@ -0,0 +1,159 @@
+_LEGACY_FUNCTION_ENV_VARS = [_FUNCTION_NAME, _FUNCTION_REGION, _FUNCTION_ENTRY]
+"""Environment variables set in Cloud Functions environments."""
+
+
+_REGION_ID = "instance/region"
+_ZONE_ID = "instance/zone"
+_GCE_INSTANCE_ID = "instance/id"
+"""Attribute in metadata server for compute region and instance."""
+
+_GKE_CLUSTER_NAME = "instance/attributes/cluster-name"
+"""Attribute in metadata server when in GKE environment."""
+
+
+def _create_functions_resource(project):
+    region = retrieve_metadata_server(_REGION_ID)
+    if _FUNCTION_NAME in os.environ:
+        function_name = os.environ.get(_FUNCTION_NAME)
+    elif _CLOUD_RUN_SERVICE_ID in os.environ:
+        function_name = os.environ.get(_CLOUD_RUN_SERVICE_ID)
+    else:
+        function_name = ""
+    resource = Resource(
+        type="cloud_function",
+        labels={
+            "project_id": project,
+            "function_name": function_name,
+            "region": region if region else "",
+        },
+    )
+    return resource
+
+
+def _create_kubernetes_resource(project):
+    zone = retrieve_metadata_server(_ZONE_ID)
+    cluster_name = retrieve_metadata_server(_GKE_CLUSTER_NAME)
+
+    resource = Resource(
+        type="k8s_container",
+        labels={
+            "project_id": project,
+            "location": zone if zone else "",
+            "cluster_name": cluster_name if cluster_name else "",
+        },
+    )
+    return resource
+
+
+def _create_compute_resource(project):
+    instance = retrieve_metadata_server(_GCE_INSTANCE_ID)
+    zone = retrieve_metadata_server(_ZONE_ID)
+    resource = Resource(
+        type="gce_instance",
+        labels={
+            "project_id": project,
+            "instance_id": instance if instance else "",
+            "zone": zone if zone else "",
+        },
+    )
+    return resource
+
+
+def _create_cloud_run_resource(project):
+    region = retrieve_metadata_server(_REGION_ID)
+    resource = Resource(
+        type="cloud_run_revision",
+        labels={
+            "project_id": project,
+            "service_name": os.environ.get(_CLOUD_RUN_SERVICE_ID, ""),
+            "revision_name": os.environ.get(_CLOUD_RUN_REVISION_ID, ""),
+            "location": region if region else "",
+            "configuration_name": os.environ.get(_CLOUD_RUN_CONFIGURATION_ID, ""),
+        },
+    )
+    return resource
+
+
+def _create_app_engine_resource(project):
+    zone = retrieve_metadata_server(_ZONE_ID)
+    resource = Resource(
+        type="gae_app",
+        labels={
+            "project_id": project,
+            "module_id": os.environ.get(_GAE_SERVICE_ENV, ""),
+            "version_id": os.environ.get(_GAE_VERSION_ENV, ""),
+            "zone": zone if zone else "",
+        },
+    )
+    return resource
+
+
+def _create_global_resource(project):
+    return Resource(type="global", labels={"project_id": project})
+
+
+def detect_resource(project):
```

Comment:
Wow this looks like quite the endeavor. Looks like it'll be very useful for users! 👏 

If you haven't already, perhaps check in with some serverless folks to see if the environment variables you've chosen are what they recommend? I know some Functions variables changed between the [3.7 and 3.8](https://docs.google.com/document/d/19o3xTj89oKrWEJFD3Gm_Ep8PReLWt2yiUGEuC09oZgM/edit#heading=h.bl5oepgz1bx5) runtimes for instance:

Reply:
Yeah good point. I've been going off [the docs](https://cloud.google.com/functions/docs/env-var), but I'll check with the team too

Reply:
I talked with other DPEs, and this seems to be in-line with how others are doing environment detection